README.md

```diff
@@ -102,6 +102,160 @@ Now re-run the "db-migrate --env dev up" command.
+
+To let business object layer decoupled with special technology platform, also, do not bring things like SQL into the services and models.
+
+### Something more about the Model layer
+
+Because we are not using ORM (Object Relationship Mapping) and not using object DB like Mongo, it's a bit challenge to build a Model system. Because we can not build models/objects like: a wallet, with the properties it has (like: name, create time), and assign methods to it which as a wallet class they should have, like: `wallet.transfer(token, anotherWallet)`, just like a classic model object looks like in OOP world, having its state, and the methods it should have.
+
+Thanks Knex, we can use it to easily retrieve objects from SQL DB, but they are simple value object, not model.
+
+So the trade-off way we are using is building model object JUST with the identity (e.g. primary key), but don't have any other properties, if we want to visit them, require the DB(repository) at the moment, the reason that we don't cache the properties value is because it's too hard to maintain the state of object (sync with the DB), this might lead us to build a kinda ORM.
+
+So the things is: create model object with identity, as a Class, they have methods to do stuff, and request DB when we need to visit the state/data which was stored in DB.
+
+### About Class vs literal object
+
+```
+class SomeService{
+  doSomething(){
+  }
+}
+```
+
+VS
+
+```
+someService = {
+  doSomething: () => {}
+}
+```
+
+There are two way to compose object, Class or direct literal object in Javascript, this project we suggest use Class to build object, like: model, service, repository, even though they are stateless objects. This is because generally to say, Class brings more flexibility for future choices, all the things literal object and do, Class can do it too, but literal object can not do all the things Class can do. 
+
+One things should be considered in future is the database transaction, if we use Class and create new instance ever time, then it's possible for us to pass the transition session object into the object to do things in a database transaction session.
+
+# Exception & Error handling
+
+We extended the library: [express-async-handler](https://github.com/Abazhenov/express-async-handler#readme) to build our customized error handling mechanism.
+
+If you need to break the normal logic work flow and inform the clients something, say, some violation according to rules, all you need to do is just throw the HttpError object:
+
+```
+throw new HttpError(403, 'Do not have permission to do this operation...');
+```
+
+The protocol layer would catch the object and return to client with this response:
+
+* Status code: 403
+
+* Response body: 
+
+```
+{
+  code: 403,
+  message: 'Do not have permission to do this operation',
+}
+```
+
+# Schema Validation
+
+We use [joi](https://joi.dev) to check JSON schema, like the input parameters for API http request.
```

Comment:
Why do we use joi to check input parameters instead of express-validator on the protocol layer? Any particular advantages here?

Reply:
@mzyau I remember we discussed this last meeting? Using generic validator to replace express-dedicated validator, because we can use the generic validator in other cases, for example, the validator a response for some other API/microservice, the advantage is we don't need to learn twice.
And I also think now the way we use joi for express validation looks easier then express-validator, just code in one place, that's all, before it seems we need to add code at two places.